Firestore/core/test/unit/core/target_test.cc

```diff
@@ -0,0 +1,361 @@
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/core/target.h"
+
+#include <cmath>
+
+#include "Firestore/core/src/core/bound.h"
+#include "Firestore/core/src/core/query.h"
+#include "Firestore/core/src/model/document_set.h"
+#include "Firestore/core/src/model/field_path.h"
+#include "Firestore/core/src/model/mutable_document.h"
+#include "Firestore/core/src/model/value_util.h"
+#include "Firestore/core/src/nanopb/message.h"
+#include "Firestore/core/test/unit/testutil/testutil.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace core {
+
+namespace {
+
+using core::Bound;
+using firebase::firestore::util::ComparisonResult;
+using model::CanonicalId;
+using model::DocumentComparator;
+using model::Equals;
+using model::FieldIndex;
+using model::FieldPath;
+using model::MutableDocument;
+using model::Segment;
+using nanopb::MakeSharedMessage;
+
+using testing::AssertionResult;
+using testing::Not;
+using testutil::Array;
+using testutil::BlobValue;
+using testutil::CollectionGroupQuery;
+using testutil::DbId;
+using testutil::Doc;
+using testutil::Field;
+using testutil::Filter;
+using testutil::MakeFieldIndex;
+using testutil::Map;
+using testutil::OrderBy;
+using testutil::Query;
+using testutil::Ref;
+using testutil::Value;
+using testutil::Vector;
+
+void VerifyBound(const absl::optional<IndexedBoundValues>& bound,
+                 bool inclusive,
+                 const std::vector<google_firestore_v1_Value>& values) {
+  ASSERT_TRUE(bound.has_value());
+  EXPECT_EQ(inclusive, bound.value().is_inclusive);
+  EXPECT_EQ(values.size(), bound.value().values.size());
+  for (size_t i = 0; i < values.size(); ++i) {
+    const auto& expected_value = values[i];
+    EXPECT_TRUE(Equals(expected_value, bound.value().values[i]))
+        << "Values should be equal: Expected: " << CanonicalId(expected_value)
+        << ", Actual: " << CanonicalId(bound.value().values[i]) << "";
+  }
+}
+
+TEST(TargetTest, EmptyQueryBound) {
+  Target target = testutil::Query("c").ToTarget();
+  FieldIndex index = MakeFieldIndex("c");
+
+  auto lower_bound = target.GetLowerBound(index);
+  VerifyBound(lower_bound.value(), true, {});
+
+  auto upper_bound = target.GetUpperBound(index);
+  VerifyBound(upper_bound.value(), true, {});
+}
+
+TEST(TargetTest, EqualsQueryBound) {
+  Target target =
+      Query("c").AddingFilter(Filter("foo", "==", "bar")).ToTarget();
+  FieldIndex index = MakeFieldIndex("c", "foo", Segment::Kind::kAscending);
+
+  auto lower_bound = target.GetLowerBound(index);
+  VerifyBound(lower_bound.value(), true, {*Value("bar")});
```

Comment:
This should just be `lower_bound`. Right now, this is converted back to an absl::optional implicitly. Applies everywhere.